runtime/src/accounts_db.rs

```diff
@@ -2259,28 +2289,279 @@ impl AccountsDb {
+                    }
+                }
+                LoadedAccountAccessor::Stored(None) => {
+                    if is_root_fixed {
+                        // When running replay on the validator, or banking stage on the leader,
+                        // it should be very rare that the storage entry doesn't exist if the
+                        // entry in the accounts index is the latest version of this account.
+                        //
+                        // There are only a few places where the storage entry may not exist
+                        // after reading the index:
+                        // 1) Shrink has removed the old storage entry and rewritten to
+                        // a newer storage entry
+                        // 2) The `pubkey` asked for in this function is a zero-lamport account,
+                        // and the storage entry holding this account qualified for zero-lamport clean.
+                        //
+                        // In both these cases, it should be safe to retry and recheck the accounts
+                        // index indefinitely, without incrementing num_acceptable_failed_iterations.
+                        // That's because if the root is fixed, there should be a bounded number
+                        // of pending cleans/shrinks (depends how far behind the AccountsBackgroundService
+                        // is), termination to the desired condition is guaranteed.
+                        //
+                        // Also note that in both cases, if we do find the storage entry,
+                        // we can guarantee that the storage entry is safe to read from because
+                        // we grabbed a reference to the storage entry while it was still in the
+                        // storage map. This means even if the storage entry is removed from the storage
+                        // map after we grabbed the storage entry, the recycler should not reset the
+                        // storage entry until we drop the reference to the storage entry.
+                        //
+                        // eh, no code in this branch? yes!
+                    } else {
+                        // RPC get_account() may have fetched an old root from the index that was
+                        // either:
+                        // 1) Cleaned up by clean_accounts(), so the accounts index has been updated
+                        // and the storage entries have been removed.
+                        // 2) Dropped by purge_slots() because the slot was on a minor fork, which
+                        // removes the slots' storage entries but doesn't purge from the accounts index
+                        // (account index cleanup is left to clean for stored slots).
+                        num_acceptable_failed_iterations += 1;
+                    }
+                }
+            }
+            if num_acceptable_failed_iterations >= 10 {
```

Comment:
I moved this check from the top of this `loop` to this place because otherwise this `datapoint_warn!` would print after updating the `(slot, store_id, offset)` tuple... much less helpful.